### ISSUE TITLE
Why do we need to assign the label value when we test a trained model?

### ISSUE THREAD

**Issue by luzhi** [vlfeat/matconvnet#62]
I found that in **cnn_train()** there is a line **net.layers{end}.class = labels ;** in line 139, which is to assign the label (ground truth) to the **net** before test an input image with the model. I tried to write my test code. If there is no such line, there would be an error in the function **vl_nnsoftmaxloss**. Could you please advise me the reason to do this? Thanks!

**Comment by lenck** [comment]
Hi, this is a way how you input the ground truth labels in the current implementation (as the network has only single input as it has a chain structure). Of course you can change it, and even more we sort of encourage to do so, but you can also easily get rid of it :)
For example, if you don't want to compute the loss, but just the features (or classification scores), you can simply replace the last layer with softmax, like:
```
net.layers{end}.type = 'softmax';
```
Then you don't need to pass the class labels anymore...

**Comment by luzhi** [comment]
Hi @lenck , thanks for your reply! Actually, I'm not sure if my code used to test an image witht the trained model is correct or not. Could you please have a look at that? The CNN structure comes from ImageNet CNN's. Thanks!

The test code is:

> net.layers{end}.class = labels;
> res = vl_simplenn(net,data);
> predicts = squeeze(res(end).x);
> if predicts > THRESHOLD
>    predicts = 1;
> else
>    predicts = 2;
> end

where the **data** is an image of  227x227x3 and **labels** is either 1 for positive. or 2 for negatives.

**Comment by lenck** [comment]
I don't think this would work... The loss, which you are comparing there is actually high for misclassification and low for correctly classified samples and also is computed as a sum over all samples (so the predicts would have always size of 1).
Better would be to do something like this:
```
net.layers{end}.type = 'softmax';
res = vl_simplenn(net,data);
[probs, predicts] = sort(squeeze(res(end).x), 1, 'descend');
```
Assuming that predicts then would be of size [NUM_CLASSES, NUM_SAMPLES] and would contain the predicted class number. The variable probs, in the first row, would contain a probability of that class...

**Comment by luzhi** [comment]
@lenck Thanks!
